### ISSUE TITLE
How do resource types / traits parameter work with libraries?

### ISSUE THREAD

**Issue by sichvoge** [raml-org/raml-spec#481]
I came across a problem when working with libraries and resource type / trait parameters. Let's have a look at the following example:

rt-lib.raml

```yaml
#%RAML 1.0 Library

resourceTypes:
  readOnlyCollection:
    description: Collection of available <<resourcePathName>> in Jukebox.
    get:
      description: Get a list of <<resourcePathName>>.
      responses:
        200:
          body:
            application/json:
              type: <<resourcePathName>>
```

types.raml

```yaml
#%RAML 1.0 Library

types:
  albums:
```

api.raml

```yaml
#%RAML 1.0 
title: Test

uses:
  rt: rt-lib.raml

/albums:
  type: rt.readOnlyCollection
```
  
As you can see in the example above, the resource type `readOnlyCollection` defines the response body type to be dynamic using `<<resourcePathName>>`. The problem here is that `<<resourcePathName>>` gets resolved into `albums` on the `/albums` resource, but that type is declared in a library which also needs to brought in using a namespace. Even if you bring that in, how can you make sure that `<<resourcePathName>>` plus the validator finds the correct type in the correct library?

**Comment by petrochenko-pavel-a** [comment]
Basically we have the following problem:

Specification threats `traits` and `resourceTypes` as macro definitions that are applied using expansion process which is documented in [Algorithm of merging traits and methods](https://github.com/raml-org/raml-spec/blob/raml-10/versions/raml-10/raml-10.md#algorithm-of-merging-traits-and-methods)

This algorithm operates in terms of nodes and properties and is not aware about namespaces and libraries. So if you in resource type definition you are referring to the type declared in the context of library, then resulting expanded API may have reference to the type name which is not valid in the global context. As for me there are following possible solutions to this:

a) It is as it is. In this case you can not actually refer to user defined types from traits/resource types declared in libraries.

Drawbacks: it really limits applicability of resource types/traits defined in libraries.

b) Clarify that when validator validates API and it validates resources/methods it should handle nodes coming from traits/resource types in a special way, and keep track about type and namespaces declared in original library from where it comes from. Currently we do not have any normative documentation on RAML file validation process. So we may handle this without doing any actual changes to the specification itself. However it definitely raises urgency for creation and publication of such document. We also need to handle this case in normative documentation for printing standard representation of RAML in JSON.

Drawbacks: Expanded version of API is not valid on its own. 

c) Make the merging algorithm aware about namespaces/libraries. In this case referenced type should be literally pulled in the context of expanded API with some unique name as well as all it's dependencies.

Drawbacks: We need to change specification, expansion process needs to gather types, track possible cycles assign unique names. 

As for me all options have some drawbacks, at the current time frame I would probably suggest to follow to the b)

Regards,
Pavel.

**Comment by petrochenko-pavel-a** [comment]
Minor correction to example:
```raml
#%RAML 1.0 Library
types:
   album:
      type: object
resourceTypes:
  readOnlyCollection:
    description: Collection of available <<resourcePathName>> in Jukebox.
    get:
      description: Get a list of <<resourcePathName>>.
      responses:
        200:
          body:
            application/json:
              type: <<resourcePathName>>
```

**Comment by machaval** [comment]
@petrochenko-pavel-a This example should work.

**Comment by fededonna** [comment]
@petrochenko-pavel-a @sichvoge this should also work.

This was the case we reviews with @machaval and @svacas 

rt-lib.yaml
``` yaml
#%RAML 1.0 Library

uses:
  t: types.raml

resourceTypes:
  readOnlyCollection:
    description: Collection of available <<resourcePathName>> in Jukebox.
    get:
      description: Get a list of <<resourcePathName>>.
      responses:
        200:
          body:
            application/json:
              type: t.<<resourcePathName>>
```


``` yaml
#%RAML 1.0 Library

types:
  albums:
```

``` yaml
#%RAML 1.0 
title: Test

uses:
  rt: rt-lib.raml

/albums:
  type: rt.readOnlyCollection
```

**Comment by sichvoge** [comment]
@fededonna @machaval @petrochenko-pavel-a did we reach any conclusion already? I see that Pavel made some suggestion which all sound reasonable, but what are you guys thinking about them?

@petrochenko-pavel-a between b) and c); I see changing the specification to add more clarification not a s a drawback but we should clarify that probably today to not mess around with GA. Whats the really drawback on gathering all the different types for c)? Is this similar to what @fededonna proposed?

**Comment by petrochenko-pavel-a** [comment]
> Whats the real drawback on gathering all the different types for c)? Is this similar to what @fededonna proposed?

Almost to best of my understanding they are doing c) in memory and using b) for final output, but I can not check it because public version of code does not contains it.

>@petrochenko-pavel-a can you also explain the drawback on b). Not sure if I can follow you here.
Primary drawback of b) is that in this case we need to explicitly declare the state named expanded version. In this case if you will serialise expanded version of API after all merging was performed and then will validate it is as a normal RAML there will be situations when it will not pass validation.

>With regards to the examples, so we have two possibilities; either we need to specify the information in the same library, or using uses if they are in a different library. Correct?.
We actually can support some special handling for resolving types from top level scope (in any of b and c), but we need to decide if we actually want to do it or not.

Regards,
Pavel

**Comment by sichvoge** [comment]
> there will be situations when it will not pass validation

@petrochenko-pavel-a can you elaborate a bit more.

**Comment by petrochenko-pavel-a** [comment]
yep. Lets look on the sample above: After expansion API will look like:

```raml
#%RAML 1.0 
title: Test
uses:
  rt: rt-lib.raml
/albums:
   description: Collection of available <<resourcePathName>> in Jukebox.
   type:  rt.readOnlyCollection
    get:
      description: Get a list of <<resourcePathName>>.
      responses:
        200:
          body:
            application/json:
              type: t.albums
```
So type `t.albums` will not be resolved and expanded API will not pass validation on its own without knowledge that API already was expanded and `get` method  is substituted from rt.readOnlyCollection

**Comment by sichvoge** [comment]
Understand; since the context is there inside the library, can we not bring in the reference as well? That might have limitations though if you have a lib defined with namespace `t` already.
